lumen/ai/agents.py

```diff
@@ -566,18 +566,25 @@ async def _create_valid_sql(
         errors=None
     ):
         if errors:
-            last_query = self.interface.serialize()[-1]["content"]
-            sql_code_match = re.search(r'(?s)```sql\s*(.*?)```', last_query, re.DOTALL)
+            last_content = self.interface.objects[-1].object[-1][-1].object
+            sql_code_match = re.search(r'(?s)```sql\s*(.*?)```', last_content, re.DOTALL)
+            chain_of_thought = last_content.split("```")[0]
             if sql_code_match:
                 last_query = sql_code_match.group(1)
-            errors = '\n'.join(errors)
-            system += (
-                f"\n\nYour last query did not work as intended:\n```sql\n{last_query}\n```\n\n"
-                f"Your priority is to expertly revise these errors:\n\n```\n{errors}\n```\n\n"
-                f"If the error is `syntax error at or near \")\"`, double check you used "
+            else:
+                return
+            num_errors = len(errors)
+            errors = ('\n'.join(f"{i+1}. {error}" for i, error in enumerate(errors))).strip()
+            content = (
+                f"\n\nYou are a world-class SQL user. Identify why this query failed:\n```sql\n{last_query}\n```\n\n"
+                f"Your goal is to try a different query to address the question while avoiding these issues:\n```\n{errors}\n```\n\n"
+                f"Please build upon your previous thought: {chain_of_thought!r}, but note, a penalty of $100 will be incurred "
+                f"for every time the issue occurs, and thus far you have been penalized ${num_errors * 100}! "
```

Comment:
Does this kind of stuff work? I've read conflicting things.

Reply:
In my testing, on the second try, all I can conclude is that it changes its query strategy.